src/lapic/lapic_msr.rs

```diff
@@ -8,31 +8,37 @@ use x86_64::registers::model_specific::Msr;
 
 use crate::lapic::LocalApicMode;
 
-
 #[derive(Debug)]
-pub struct LocalApicRegister { msr: Msr, mmio_offset: u32 }
+pub enum LocalApicRegister {
+    X2apic(Msr),
+    XapicOffset(u64),
+}
 
 impl LocalApicRegister {
-    pub fn new(register: (u32, u32)) -> Self {
-        Self { msr: Msr::new(register.0), mmio_offset: register.1 }
-    }
-
-    pub unsafe fn mmio_address(&self) -> u64 {
-        let base = Msr::new(IA32_APIC_BASE).read() & 0xFFFFFF000;
-        base + self.mmio_offset as u64
+    pub fn new(mode: LocalApicMode, loc: (u32, u32)) -> Self {
+        match mode {
+            LocalApicMode::XApic { xapic_base } => {
+                Self::XapicOffset(xapic_base + loc.1 as u64)
+            }
+            LocalApicMode::X2Apic => Self::X2apic(Msr::new(loc.0)),
+        }
     }
 
-    pub unsafe fn read(&self, mode: &LocalApicMode) -> u64 {
-        match mode {
-            LocalApicMode::XApic => *(self.mmio_address() as *const u64),
-            LocalApicMode::X2Apic => self.msr.read()
+    pub unsafe fn read(&self) -> u64 {
+        match self {
+            Self::XapicOffset(addr) => {
+                core::ptr::read_volatile(addr as *const u64)
+            }
+            Self::X2apic(msr) => msr.read(),
         }
     }
 
-    pub unsafe fn write(&mut self, mode: &LocalApicMode, value: u64) {
-        match mode {
-            LocalApicMode::XApic => { *(self.mmio_address() as *mut u64) = value },
-            LocalApicMode::X2Apic => self.msr.write(value)
+    pub unsafe fn write(&mut self, value: u64) {
+        match self {
+            LocalApicRegister::XapicOffset(offset) => {
+                *(offset as *mut u64) = value
```

Comment:
There are two subtle errors here:
1. `offset` is a reference to the field in the enum, so `offset as *mut u64` gets a pointer to this field, but really we wanted to cast the value of the field to a pointer. The correct version would be `*offset as *mut u64`
2. You should use `core::ptr::write_volatile` here to make sure the read isn't optimized away or reordered

The corrected code should look like this:
`core::ptr::write_volatile(*offset as *mut u64, value)`